janus_server/src/datastore.rs

```diff
@@ -423,85 +467,66 @@ impl Transaction<'_> {
-                A::OutputShare::try_from(&vdaf_message_bytes.ok_or_else(|| {
-                    Error::DbState(
-                        "report aggregation in state FINISHED but vdaf_message is NULL".to_string(),
-                    )
-                })?)
-                .map_err(|err| {
-                    Error::DecodeError(CodecError::Other(
-                        format!("couldn't decode output share: {}", err).into(),
-                    ))
-                })?,
-            ),
-            ReportAggregationStateCode::Failed => {
-                ReportAggregationState::Failed(error_code.ok_or_else(|| {
-                    Error::DbState(
-                        "report aggregation in state FAILED but error_code is NULL".to_string(),
-                    )
-                })?)
-            }
-            ReportAggregationStateCode::Invalid => ReportAggregationState::Invalid(),
-        };
+        self.tx
+            .query(
+                &stmt,
+                &[
+                    /* aggregation_job_id */ &&aggregation_job_id.0[..],
+                    /* task_id */ &&task_id.0[..],
+                    /* nonce_time */ &nonce_time,
+                    /* nonce_rand */ &nonce_rand,
+                ],
+            )
+            .await
+            .map_err(Error::from)
+            .and_then(single_row)
+            .and_then(|row| {
+                report_aggregation_from_row(verify_param, task_id, aggregation_job_id, row)
+            })
+    }
 
-        Ok(ReportAggregation {
-            aggregation_job_id,
-            task_id,
-            nonce,
-            ord,
-            state: agg_state,
-        })
+    /// get_report_aggregations_for_aggregation_job retrieves all report aggregations associated
+    /// with a given aggregation job, ordered by their natural ordering.
+    pub async fn get_report_aggregations_for_aggregation_job<A: vdaf::Aggregator>(
+        &self,
+        verify_param: &A::VerifyParam,
+        task_id: TaskId,
+        aggregation_job_id: AggregationJobId,
+    ) -> Result<Vec<ReportAggregation<A>>, Error>
+    where
+        A::PrepareStep: ParameterizedDecode<A::VerifyParam>,
+        A::OutputShare: for<'a> TryFrom<&'a [u8]>,
+        for<'a> &'a A::AggregateShare: Into<Vec<u8>>,
+    {
+        let stmt = self
+            .tx
+            .prepare_cached(
+                "SELECT client_reports.nonce_time, client_reports.nonce_rand, report_aggregations.ord, report_aggregations.state, report_aggregations.vdaf_message, report_aggregations.error_code
+                FROM report_aggregations
+                JOIN client_reports ON client_reports.id = report_aggregations.client_report_id
+                WHERE report_aggregations.aggregation_job_id = (SELECT id FROM aggregation_jobs WHERE aggregation_job_id = $1)
+                  AND client_reports.task_id = (SELECT id FROM tasks WHERE task_id = $2)"
+            )
```

Comment:
Oof, nice catch -- idk how I didn't catch this with testing (but I suspect you're correct about the tests not being "big" enough to trigger whatever behavior internal to postgres that would cause the rows to be returned in non-insertion order).